### ISSUE TITLE
F601 `multi-value-repeated-key-literal` various issues

### ISSUE THREAD

**Issue by dylwil3** [astral-sh/ruff#12911]
This is an extension of #12772.

# Observed issues
[Playground Link](https://play.ruff.rs/e6b35fd7-dca9-421a-915a-fa3d34f04dd9)

- F601 should only be triggered for Expr's whose leaves are literals of some kind - the following two examples ought to fall under F602 instead.

```python
{(foo, bar): "a", (foo, bar): "b"}

{
    f"{foo}": "a",
    f"{foo}": "b",
}
```

- As in https://github.com/astral-sh/ruff/issues/12772 booleans and numbers that are equal should trigger this rule.

```python
{
    True: "a",
    1: "b",
    1.0: "c",
    1.0
    + 0.0j: "d",  # <-- this may be out of scope depending on the implementation of the fix
    2: "e",
    2.0: "f",
}
```

- (Maybe a feature request rather than a bug) It would be nice to extend F601 to situations like this, where expression ASTs have _literal_ leaves and internal nodes do not involve a function call

```python
{
    -1: "a",
    -1: "b",
}

{
    1 + 1: "a",
    1 + 1: "b",
}

# but _not_ to situations which require any kind of actual evaluation
{
    1 + 1: "a",
    2: "b",
}

# except maybe this, which appears above as well
# and which is arguably a casting/type promotion
{
    1: "a",
    1.0 + 0.0j: "b",
}
```

- There is also an issue with the suggested fix for values which are the "same", as there could be side effects, eg

```python
# Suggested fix is unsafe, since we can't know
# which value of iterator was desired
{
    1: next(it),
    1: next(it),
}
```

- Finally, a minor issue: the name of the rule does not seem quite right, since the linter complains if a key is repeated even if the same value appears.

# Potential Solution

I propose that the _logic_ of this check (independently of the implementation) should be as follows:

1. F601 should ignore keys which are not:
    - Expressions that have leaves only literals 
    - Expressions where no node is a `Name`
2. F601 should consider keys duplicated if the corresponding `ComparableExpr`'s are equal after further replacing all boolean and numeric literals with their complex-valued equivalent (for these purposes, I think "large" integers should be left as is).
3. F601 should not propose a fix if the "unique" value found has possible side effects - for example, we could just never propose a fix if the value found involves a function call.
4. Change the name of the rule to `repeated-key-literal` (and similarly for F602).

**Comment by dylwil3** [comment]
By the way, I'm happy to work on this I just wanna pin down the "assignment" first.

**Comment by MichaReiser** [comment]
I hope you don't mind if I convert this to a discussion.
